staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go

```diff
@@ -503,7 +504,14 @@ func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
 		// wait for the delayed stopCh before closing the handler chain (it rejects everything after Wait has been called).
 		<-delayedStopCh.Signaled()
 
+		// From https://github.com/kubernetes/kubernetes/pull/110000
+		// Additionally wait for preshutdown hooks to also be finished, as some of them need
+		// to send API calls to clean up after themselves (e.g. lease reconcilers removing
+		// itself from the active servers).
+		<-preShutdownHooksHasStoppedCh.Signaled()
```

Comment:
If I remove this, then the `TestGracefulTerminationWithKeepListeningDuringGracefulTerminationDisabled` fails with the following error:
```
$ go test -race -v k8s.io/apiserver/pkg/server -run=TestGracefulTerminationWithKeepListeningDuringGracefulTerminationDisabled -args -v=1
=== RUN   TestGracefulTerminationWithKeepListeningDuringGracefulTerminationDisabled
    genericapiserver_graceful_termination_test.go:595: Sending request - timeout: 100ms, url: https://127.0.0.1:38697/echo?message=attempt-1
I0512 20:11:56.640675 4143338 genericapiserver.go:445] MuxAndDiscoveryComplete has all endpoints registered and discovery information is complete
I0512 20:11:56.642451 4143338 secure_serving.go:210] Serving securely on [::]:38697
I0512 20:11:56.642555 4143338 tlsconfig.go:240] "Starting DynamicServingCertificateController"
    genericapiserver_graceful_termination_test.go:706: [server] seen new connection: &net.TCPConn{conn:net.conn{fd:(*net.netFD)(0xc00078e000)}}
I0512 20:11:56.643440 4143338 genericapiserver.go:518] [graceful-termination] waiting for shutdown to be initiated
    genericapiserver_graceful_termination_test.go:688: [server] received a request, proto: HTTP/2.0, url: /echo?message=attempt-1
    genericapiserver_graceful_termination_test.go:578: The API server has started
    genericapiserver_graceful_termination_test.go:595: Sending request - timeout: 0s, url: https://127.0.0.1:38697/in-flight-request-as-designed
    genericapiserver_graceful_termination_test.go:706: [server] seen new connection: &net.TCPConn{conn:net.conn{fd:(*net.netFD)(0xc0006e6000)}}
I0512 20:11:56.675658 4143338 genericapiserver.go:458] "[graceful-termination] shutdown event" name="ShutdownInitiated"
I0512 20:11:56.675744 4143338 object_count_tracker.go:84] "StorageObjectCountTracker pruner is exiting"
I0512 20:11:56.776444 4143338 genericapiserver.go:461] "[graceful-termination] shutdown event" name="AfterShutdownDelayDuration"
    genericapiserver_graceful_termination_test.go:166: Before ShutdownDelayDuration elapses new request(s) should be served
    genericapiserver_graceful_termination_test.go:595: Sending request - timeout: 1s, url: https://127.0.0.1:38697/echo?message=request-on-an-existing-connection-should-succeed
    genericapiserver_graceful_termination_test.go:688: [server] received a request, proto: HTTP/2.0, url: /echo?message=request-on-an-existing-connection-should-succeed
    genericapiserver_graceful_termination_test.go:595: Sending request - timeout: 1s, url: https://127.0.0.1:38697/echo?message=request-on-a-new-tcp-connection-should-succeed
    genericapiserver_graceful_termination_test.go:706: [server] seen new connection: &net.TCPConn{conn:net.conn{fd:(*net.netFD)(0xc0003ab180)}}
    genericapiserver_graceful_termination_test.go:688: [server] received a request, proto: HTTP/2.0, url: /echo?message=request-on-a-new-tcp-connection-should-succeed
I0512 20:11:56.811131 4143338 genericapiserver.go:514] "[graceful-termination] will reject incoming request(s) here on" name="HandlerChainWaitGroup.Wait()"
    genericapiserver_graceful_termination_test.go:595: Sending request - timeout: 1s, url: https://127.0.0.1:38697/echo?message=request-from-pre-shutdown-hook-should-succeed
    genericapiserver_graceful_termination_test.go:706: [server] seen new connection: &net.TCPConn{conn:net.conn{fd:(*net.netFD)(0xc00078e200)}}
    genericapiserver_graceful_termination_test.go:544: Expected Status Code: 200, but got: 503
I0512 20:11:56.835659 4143338 genericapiserver.go:526] "[graceful-termination] pre-shutdown hooks completed" name="PreShutdownHooksCompleted"
I0512 20:11:56.835736 4143338 tlsconfig.go:255] "Shutting down DynamicServingCertificateController"
I0512 20:11:56.835941 4143338 secure_serving.go:255] Stopped listening on [::]:38697
I0512 20:11:56.836041 4143338 genericapiserver.go:497] "[graceful-termination] shutdown event" name="HTTPServerStoppedListening"
I0512 20:11:56.836105 4143338 genericapiserver.go:516] "[graceful-termination] shutdown event" name="InFlightRequestsDrained"
    genericapiserver_graceful_termination_test.go:254: Waiting for the apiserver Run method to return
I0512 20:11:57.913447 4143338 genericapiserver.go:539] [graceful-termination] apiserver is exiting
--- FAIL: TestGracefulTerminationWithKeepListeningDuringGracefulTerminationDisabled (1.28s)
FAIL
FAIL	k8s.io/apiserver/pkg/server	1.392s
FAIL

```

Reply:
:clap: